sql: allow FK over stored computed columns

spilchen opened the pull request:
Previously, any attempt to construct a foreign key that used a computed column was blocked. This will remove that block as it is now a supported feature.

Even though this is about removing the blocking code, I had to add some new blocking code since we cannot support all options for foreign key. Specifically, any of the ON UPDATE or ON DELETE clauses that would cause an update to the computed column. Since the column value is computed, we cannot allow it to change. Note, we do support ON DELETE CASCADE since that doesn't modify the row rather it removes the row entirely.

There is some duplication of the checks, since it required updates in two codepaths for ALTER TABLE to handle the different schema changer. I tried to modularize as best as possible.

Fixes: #46672
Release note (sql change): Allow foreign keys to be created over computed columns. However, most ON UPDATE and ON DELETE options for foreign key constraints are not allowed with computed columns.  Only the following are supported:
ON UPDATE (NO ACTION|RESTRICT), ON DELETE (NO ACTION|RESTRICT|CASCADE).

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/124851)
<!-- Reviewable:end -->

annrpom commented:
👋 drive-by comment: we likely want to revisit some of the expected error checks (and remove them) in our random schemachange workload for foreign key constraints
https://github.com/cockroachdb/cockroach/blob/cf7ad52a113613bef9bdf9d380c9eec4d8715b85/pkg/workload/schemachange/operation_generator.go#L845-L853

spilchen commented:
I overlooked that this PR currently only addresses stored computed columns. I didn't realize that virtual columns, which are also computed but not stored on disk, are another type. Should I expand the scope to include virtual columns as well?

rafiss commented:
Nice catch. Yes, let's make this PR also handle virtual columns.

spilchen commented:
I had trouble adding support for virtual columns in FK. If we end up scanning a table with a virtual column (whether it's the parent or child in the FK relationship), the scan logic encounters an internal error: scanned virtual column depends on non-scanned column.

This error occurs in a couple of scenarios:
- If the virtual column is in the parent table, inserting into the child table causes the scan of the parent table to validate the FK to fail.
- If the virtual column is in the child table, updating the parent table with an ON UPDATE/DELETE clause in the FK causes the scan of the child table to fail.

I tried to investigate this issue but didn't make much progress. For this PR, it makes sense to limit the scope to stored computed columns. There is a separate issue specifically for supporting virtual columns in FKs (issue #59671).

spilchen commented:
> 👋 drive-by comment: we likely want to revisit some of the expected error checks (and remove them) in our random schemachange workload for foreign key constraints
> 
> https://github.com/cockroachdb/cockroach/blob/cf7ad52a113613bef9bdf9d380c9eec4d8715b85/pkg/workload/schemachange/operation_generator.go#L845-L853

Thanks Annie. This was a good catch. I added some changes to this workload.

rafiss commented:
> I tried to investigate this issue but didn't make much progress. For this PR, it makes sense to limit the scope to stored computed columns. There is a separate issue specifically for supporting virtual columns in FKs (issue https://github.com/cockroachdb/cockroach/issues/59671).

Thanks for investigating this more! Yes I agree then. Let's leave this PR just focused on stored columns. You can work on https://github.com/cockroachdb/cockroach/issues/59671 as well separately, but I suggest time-boxing your effort, since it doesn't feel worth prioritizing that work if it's going to be a large amount of work.

(Reviewing this now)

rafiss reviewed (changes requested):
this is looking great. i left some minor comments, and after those this should be all set.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/124851#-:-Nz_5Mp9A5vkvp_xG7-u:b9mrkk7)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @DarrylWong, @herkolategan, and @spilchen)

___
*[`pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table_add_constraint.go` line 811 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/124851#-Nz_3ZtE9_UrOj_eDS25:-Nz_3ZtE9_UrOj_eDS26:bmsj7vr) ([raw file](https://github.com/cockroachdb/cockroach/blob/d97260e0ceb72da61f56e8d31642d8b1425bb728/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table_add_constraint.go#L811)):*
> ```Go
> 	// actions that would try to change the computed value. Computed values cannot
> 	// be altered directly, so attempts to set them to NULL or a DEFAULT value are
> 	// blocked.
> ```

thanks for adding good comments throughout this PR! makes it very easy to follow the changes.
___
*[`pkg/sql/sqlerrors/errors.go` line 209 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/124851#-Nz_1w1Y68Yv-JrV9wpe:-Nz_1w1Y68Yv-JrV9wpf:b-hgqa8v) ([raw file](https://github.com/cockroachdb/cockroach/blob/d97260e0ceb72da61f56e8d31642d8b1425bb728/pkg/sql/sqlerrors/errors.go#L209)):*
> ```Go
> 	// Pick the 'ON UPDATE' or 'ON DELETE' keyword. If both actions are set we
> 	// include 'ON UPDATE' in the error text. This is consistent with postgres.
> 	keyword := "DELETE"
> ```

nit: let's make the `keyword` variable be of the type `redact.SafeString`. this type is an alias for `string`, but it has a special property that it will not get redacted when errors are logged or reported in Splunk/Datadog. by default, we need to consider all values of type `string` as "unsafe" -- meaning they potentially can contain sensitive customer data so must be redacted. but in cases like this, when we ourselves are the ones crafting the string, we can opt-out of the redaction by using `redact.SafeString`.

more details are described here: https://cockroachlabs.atlassian.net/wiki/spaces/CRDB/pages/1824817806/Log+and+error+redactability
___
*[`pkg/sql/logictest/testdata/logic_test/alter_table` line 3906 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/124851#-Nz_2rd13Nh59wivNC-r:-Nz_2rd13Nh59wivNC-s:bqh4ulb) ([raw file](https://github.com/cockroachdb/cockroach/blob/d97260e0ceb72da61f56e8d31642d8b1425bb728/pkg/sql/logictest/testdata/logic_test/alter_table#L3906)):*
> ```
> 
> statement error pq: invalid ON UPDATE action for foreign key constraint containing computed column
> ALTER TABLE y ADD FOREIGN KEY (computed_col) REFERENCES x ON UPDATE CASCADE;
> ```

nit: for the sake of getting full coverage of legacy and declarative schema changer behavior, let's separate these into separate statements. the behavior (as part of postgres compatibility), is that when multiple statements are sent in one batch, the entire batch is executed in one implicit transaction. (see https://www.postgresql.org/docs/current/protocol-flow.html#PROTOCOL-FLOW-MULTI-STATEMENT).

and due to our current limitations, whenever a transaction (either implicit or explicit) has multiple schema changes, we always fall back to using the legacy schema changer. so we can avoid this by separating each statement into it's own line.

after doing that, we get coverage of both schema changers because logic tests are run under multiple "configurations." by default, the declarative schema changer is on, but under the `local-legacy-schema-changer` configuration, the declarative schema changer is fully disabled by running `SET CLUSTER SETTING sql.defaults.use_declarative_schema_changer='off'`

https://github.com/cockroachdb/cockroach/blob/0292b4b20f48029857fc6e10f7f174be84c52a52/pkg/sql/logictest/logictestbase/logictestbase.go#L292-L297

(and i would suggest the same changes for the `computed` logic test as well)


<!-- Sent from Reviewable.io -->

spilchen committed:
sql: allow FK over stored computed columns

Previously, any attempt to construct a foreign key that used a computed
column was blocked. This will remove that block when the computed column
is stored. We continue to block for virtual computed columns.

Even though this is about removing the blocking code, I had to add some
new blocking code since we cannot support all options for foreign key.
Specifically, any of the ON UPDATE or ON DELETE clauses that would cause
an update to the computed column. Since the column value is computed, we
cannot allow it to change. Note, we do support ON DELETE CASCADE since
that doesn't modify the row rather it removes the row entirely.

There is some duplication of the checks, since it required updates in
two codepaths for ALTER TABLE to handle the different schema changer. I
tried to modularize as best as possible.

Epic: none
Fixes: #46672
Release note (sql change): Allow foreign keys to be created over
stored computed columns. However, most ON UPDATE and ON DELETE options for
foreign key constraints are not allowed with computed columns.  Only the
following are supported:
ON UPDATE (NO ACTION|RESTRICT), ON DELETE (NO ACTION|RESTRICT|CASCADE).

spilchen reviewed (commented):
tftr!

I applied your suggestions.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/124851#-:-Nzcc8hv4LD_n3OwcqCh:b-j88cka)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @DarrylWong, @herkolategan, and @rafiss)

___
*[`pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table_add_constraint.go` line 811 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/124851#-Nz_3ZtE9_UrOj_eDS25:-NzcbtKO0RSGMnJYQPk3:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d97260e0ceb72da61f56e8d31642d8b1425bb728/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table_add_constraint.go#L811)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

thanks for adding good comments throughout this PR! makes it very easy to follow the changes.
</blockquote></details>

Done.
___
*[`pkg/sql/sqlerrors/errors.go` line 209 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/124851#-Nz_1w1Y68Yv-JrV9wpe:-NzcbvyKCYUtdKk5vnpw:bfhtgtn) ([raw file](https://github.com/cockroachdb/cockroach/blob/d97260e0ceb72da61f56e8d31642d8b1425bb728/pkg/sql/sqlerrors/errors.go#L209)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

nit: let's make the `keyword` variable be of the type `redact.SafeString`. this type is an alias for `string`, but it has a special property that it will not get redacted when errors are logged or reported in Splunk/Datadog. by default, we need to consider all values of type `string` as "unsafe" -- meaning they potentially can contain sensitive customer data so must be redacted. but in cases like this, when we ourselves are the ones crafting the string, we can opt-out of the redaction by using `redact.SafeString`.

more details are described here: https://cockroachlabs.atlassian.net/wiki/spaces/CRDB/pages/1824817806/Log+and+error+redactability
</blockquote></details>

This is good to know thanks.
___
*[`pkg/sql/logictest/testdata/logic_test/alter_table` line 3906 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/124851#-Nz_2rd13Nh59wivNC-r:-Nzcc0fTC4bwPDlp3l7Z:b-bqoo57) ([raw file](https://github.com/cockroachdb/cockroach/blob/d97260e0ceb72da61f56e8d31642d8b1425bb728/pkg/sql/logictest/testdata/logic_test/alter_table#L3906)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

nit: for the sake of getting full coverage of legacy and declarative schema changer behavior, let's separate these into separate statements. the behavior (as part of postgres compatibility), is that when multiple statements are sent in one batch, the entire batch is executed in one implicit transaction. (see https://www.postgresql.org/docs/current/protocol-flow.html#PROTOCOL-FLOW-MULTI-STATEMENT).

and due to our current limitations, whenever a transaction (either implicit or explicit) has multiple schema changes, we always fall back to using the legacy schema changer. so we can avoid this by separating each statement into it's own line.

after doing that, we get coverage of both schema changers because logic tests are run under multiple "configurations." by default, the declarative schema changer is on, but under the `local-legacy-schema-changer` configuration, the declarative schema changer is fully disabled by running `SET CLUSTER SETTING sql.defaults.use_declarative_schema_changer='off'`

https://github.com/cockroachdb/cockroach/blob/0292b4b20f48029857fc6e10f7f174be84c52a52/pkg/sql/logictest/logictestbase/logictestbase.go#L292-L297

(and i would suggest the same changes for the `computed` logic test as well)
</blockquote></details>

Thanks for the background. This makes sense.


<!-- Sent from Reviewable.io -->

rafiss reviewed (approved):
lgtm! nice work

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/124851#-:-NzdA8go4fNwwA31X7Hu:b-m0nb8k)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @DarrylWong and @herkolategan)



<!-- Sent from Reviewable.io -->

spilchen commented:
bors r=rafiss
